src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -823,26 +825,53 @@ private async Task CreateServicesAsync(CancellationToken cancellationToken = def
-                if (srvResource == null) { continue; } // This service most likely already has full address information, so it is not on needAddressAllocated list.
+            for (int attempt = 0; attempt < watchAttempts; attempt++)
+            {
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                cts.CancelAfter(configuration.GetValue<TimeSpan>("DOTNET_ASPIRE_SERVICE_STARTUP_WATCH_TIMEOUT", TimeSpan.FromSeconds(10)));
 
-                if (updated.HasCompleteAddress || updated.Spec.AddressAllocationMode == AddressAllocationModes.Proxyless)
+                try
                 {
-                    srvResource.Service.ApplyAddressInfoFrom(updated);
-                    needAddressAllocated.Remove(srvResource);
+                    // We do not specify the initial list version, so the watcher will give us all updates to Service objects.
+                    IAsyncEnumerable<(WatchEventType, Service)> serviceChangeEnumerator = kubernetesService.WatchAsync<Service>(cancellationToken: cts.Token);
+                    await foreach (var (evt, updated) in serviceChangeEnumerator)
+                    {
+                        if (evt == WatchEventType.Bookmark) { continue; } // Bookmarks do not contain any data.
+
+                        var srvResource = needAddressAllocated.FirstOrDefault(sr => sr.Service.Metadata.Name == updated.Metadata.Name);
+                        if (srvResource == null) { continue; } // This service most likely already has full address information, so it is not on needAddressAllocated list.
+
+                        if (updated.HasCompleteAddress)
+                        {
+                            srvResource.Service.ApplyAddressInfoFrom(updated);
+                            needAddressAllocated.Remove(srvResource);
+                        }
+
+                        if (needAddressAllocated.Count == 0)
+                        {
+                            return; // We are done
+                        }
+                    }
                 }
+                catch { }
+            }
 
-                if (needAddressAllocated.Count == 0)
+            // If there are still services that need address allocated, try a final direct query in case the watch missed some updates.
+            foreach(var sar in needAddressAllocated)
+            {
+                var dcpSvc = await kubernetesService.GetAsync<Service>(sar.Service.Metadata.Name, cancellationToken: cancellationToken).ConfigureAwait(false);
+                if (dcpSvc.HasCompleteAddress)
+                {
+                    sar.Service.ApplyAddressInfoFrom(dcpSvc);
+                }
+                else
                 {
-                    return; // We are done
+                    distributedApplicationLogger.LogWarning($"Unable to allocate a network port for service '{sar.Service.Metadata.Name}'; service may be unreachable and its clients may not work properly.");
```

Comment:
Don't use interpolated strings

```suggestion
                    distributedApplicationLogger.LogWarning("Unable to allocate a network port for service '{ServiceName}'; service may be unreachable and its clients may not work properly.", sar.Service.Metadata.Name);
```